rumqttd/src/router/routing.rs

```diff
@@ -304,23 +303,32 @@ impl Router {
         // Retrieve previous connection state from graveyard
         let saved = self.graveyard.retrieve(&client_id);
         let clean_session = connection.clean;
-        let previous_session = saved.is_some();
+        let previous_session = saved.as_ref().is_some_and(|s| s.session_state.is_some());
         // for qos2 pending pubrels
         let mut pending_acks = VecDeque::new();
+
         let tracker = if !clean_session {
-            let saved = saved.map_or(SavedState::new(client_id.clone()), |s| s);
-            connection.subscriptions = saved.subscriptions;
-            connection.events = saved.metrics;
-            // for using in acklog
-            pending_acks = saved.unacked_pubrels.clone();
-            outgoing.unacked_pubrels = saved.unacked_pubrels;
-            saved.tracker
+            let saved_state = saved.and_then(|saved| {
+                connection.events = saved.metrics;
+                saved.session_state
+            });
+
+            saved_state.map_or_else(
+                || Tracker::new(client_id.clone()),
+                |session_state| {
+                    connection.subscriptions = session_state.subscriptions;
+                    // for using in acklog
+                    pending_acks = session_state.unacked_pubrels.clone();
+                    outgoing.unacked_pubrels = session_state.unacked_pubrels;
+                    session_state.tracker
+                },
+            )
```

Comment:
looks like its going to be hard to read this way as well, add comment on top of this explaining what it does. else LGTM.

Reply:
> looks like its going to be hard to read this way as well
>
couldn't think of any other better way :sweat_smile: 

> add comment on top of this explaining what it does
> 
Done :smile:

Thanks for the review :)